Sources/Sentry/SentryScopeSyncC.c

```diff
@@ -0,0 +1,225 @@
+    if (size == getRawBreadcrumbSize()) {
+        return;
+    }
+
+    strcat(crumbs, breadcrumbsStart);
+    for (int i = 0; i < maxCrumbs; i++) {
+        if (breadcrumbs[i] != NULL) {
+            strcat(crumbs, "{");
+            strcat(crumbs, breadcrumbs[i]);
+            strcat(crumbs, "},");
+        }
+    }
+
+    size_t length = strlen(crumbs);
+    crumbs[length - 1] = ']';
+    crumbs[length] = '\0';
+    strcat(crumbs, ",");
+
+    strcat(destination, crumbs);
+
+    free(crumbs);
+}
+
+/**
+ * We don't lock access to the properties in this method as this is called when a crash occurs.
+ */
+void
+sentryscopesync_getJSON(char **json)
+{
+    size_t brackets = 2;
+    size_t nullByte = 1;
+    size_t resultSize = getSize(userJSON) + getSize(distJSON) + getSize(contextJSON)
+        + getSize(environmentJSON) + getSize(tagsJSON) + getSize(extrasJSON)
+        + getSize(fingerprintJSON) + getSize(levelJSON) + getBreadcrumbSize() + NUMBER_OF_FIELDS
+        + brackets + nullByte;
+
+    char *result = calloc(1, resultSize);
+
+    if (resultSize == NUMBER_OF_FIELDS + brackets + nullByte + getRawBreadcrumbSize()) {
+        // All fields are empty
+        strcat(result, "{}");
+    } else {
+        strcat(result, "{");
+        add(result, userJSON);
+        add(result, distJSON);
+        add(result, contextJSON);
+        add(result, environmentJSON);
+        add(result, tagsJSON);
+        add(result, extrasJSON);
+        add(result, fingerprintJSON);
+        add(result, levelJSON);
+        addBreadcrumbs(result);
+
+        size_t length = strlen(result);
+        result[length - 1] = '}';
+        result[length] = '\0';
+    }
+
+    *json = result;
+}
+
+static void
+set(const char *const newJSON, const char **field)
+{
+    free((void *)*field);
+    if (newJSON == NULL) {
+        *field = NULL;
+    } else {
+        *field = strdup(newJSON);
+    }
+}
+
+void
+sentryscopesync_setUser(const char *const json)
+{
+    set(json, &userJSON);
+}
+
+void
+sentryscopesync_setDist(const char *const json)
+{
+    set(json, &distJSON);
+}
+
+void
+sentryscopesync_setContext(const char *const json)
+{
+    set(json, &contextJSON);
+}
+
+void
+sentryscopesync_setEnvironment(const char *const json)
+{
+    set(json, &environmentJSON);
+}
+
+void
+sentryscopesync_setTags(const char *const json)
+{
+    set(json, &tagsJSON);
+}
+
+void
+sentryscopesync_setExtras(const char *const json)
+{
+    set(json, &extrasJSON);
+}
+
+void
+sentryscopesync_setFingerprint(const char *const json)
+{
+    set(json, &fingerprintJSON);
+}
+
+void
+sentryscopesync_setLevel(const char *const json)
+{
+    set(json, &levelJSON);
+}
+
+void
+sentryscopesync_addBreadcrumb(const char *const json)
+{
+    set(json, &breadcrumbs[currentCrumb]);
+    // Ring buffer
+    currentCrumb = (currentCrumb + 1) % maxCrumbs;
+}
+
+void
+sentryscopesync_clearBreadcrumbs(void)
+{
+    for (int i = 0; i < maxCrumbs; i++) {
+        set(NULL, &breadcrumbs[i]);
```

Comment:
you never free `breadcrumbs` however. This means that you leak memory everytime you `configureBreadcrumbs`.

Reply:
`set` actually calls free. Is this still a problem then?